# Limpieza de datos

* Vamos a aplicar el [checklist del Banco Mundial](https://dimewiki.worldbank.org/wiki/Checklist:_Data_Cleaning) en Python usando el [SciPy stack](https://www.scipy.org/stackspec.html), principalmente pandas

## 0. Prerequisitos

In [1]:
# Para que la notebook corra tanto en jupyter como en colab.
# NO ES RECOMENDADO realizar pip install desde un notebook.

import sys
in_colab = 'google.colab' in sys.modules

if in_colab:
  BASE_DIR = "https://github.com/DiploDatos/AnalisisYCuracion/raw/master/"
else:
  BASE_DIR = ".."
if 'ftfy' not in sys.modules:
    !pip install 'ftfy<5.6'

El sistema no puede encontrar el archivo especificado.


# 1. Importando los datos

# 1.1. Verificar que no hay problemas en la importación

In [2]:
import pandas as pd

# para que los numeros flotantes me los muestre solo con dos decimales
pd.options.display.float_format = '{:.2f}'.format

In [3]:
# Importamos datos de proyectos de Kickstarter la plataforma de Crowdsourcing

kickstarter_2018 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv")

In [4]:
kickstarter_2018[:5]

ID                                               name  \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000003930      Greeting From Earth: ZGAC Arts Capsule For ET   
2  1000004038                                     Where is Hank?   
3  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
4  1000011046  Community Film Project: The Art of Neighborhoo...   

         category main_category currency    deadline     goal  \
0          Poetry    Publishing      GBP  2015-10-09  1000.00   
1  Narrative Film  Film & Video      USD  2017-11-01 30000.00   
2  Narrative Film  Film & Video      USD  2013-02-26 45000.00   
3           Music         Music      USD  2012-04-16  5000.00   
4    Film & Video  Film & Video      USD  2015-08-29 19500.00   

              launched  pledged     state  backers country  usd pledged  \
0  2015-08-11 12:12:28     0.00    failed        0      GB         0.00   
1  2017-09-02 04:43:57  2421.00    failed       15      US       100.00   
2  2013-01-12 00:20:50   220.00    failed        3      US       220.00   
3  2012-03-17 03:24:11     1.00    failed        1      US         1.00   
4  2015-07-04 08:35:03  1283.00  canceled       14      US      1283.00   

   usd_pledged_real  usd_goal_real  
0              0.00        1533.95  
1           2421.00       30000.00  
2            220.00       45000.00  
3              1.00        5000.00  
4           1283.00       19500.00

In [5]:
# Set seed for reproducibility

# Al fijar la semilla no habrá azar.
# Esto es algo necesario cuando queremos "reproducir valores aleatorios".

import numpy as np
np.random.seed(0)

In [6]:
kickstarter_2018.sample(5)

ID                                               name  \
338862   796196901                                 10G Christmas Tree   
277871   483825010                                              Gliff   
47000    123916947                                 STUFFED Food Truck   
111338  1565733636  NeoExodus Adventure: Origin of Man for Pathfin...   
53743   1273544891        NAPOLEON IN NEW YORK! an original TV Series   

               category main_category currency    deadline     goal  \
338862              Art           Art      USD  2010-12-26 10526.00   
277871  Gaming Hardware         Games      USD  2016-03-28 10000.00   
47000       Food Trucks          Food      USD  2015-01-06 60000.00   
111338   Tabletop Games         Games      USD  2012-05-01   500.00   
53743            Comedy  Film & Video      USD  2016-07-26 25000.00   

                   launched  pledged       state  backers country  \
338862  2010-12-08 08:44:04     0.00      failed        0      US   
277871  2016-01-28 04:56:18    51.00      failed        5      US   
47000   2014-11-07 02:24:36    25.00      failed        1      US   
111338  2012-03-15 01:16:10   585.00  successful       17      US   
53743   2016-05-27 00:07:25    25.00      failed        1      US   

        usd pledged  usd_pledged_real  usd_goal_real  
338862         0.00              0.00       10526.00  
277871        51.00             51.00       10000.00  
47000         25.00             25.00       60000.00  
111338       585.00            585.00         500.00  
53743         25.00             25.00       25000.00

* Veamos si la descripción del dataset se corresponde con lo levantado    https://www.kaggle.com/kemical/kickstarter-projects/data

In [7]:
pd.DataFrame([["ID", "No description provided", "Numeric"],
["name", "No description provided", "String"],
["category", "No description provided", "String"],
["main_category", "No description provided", "String"],
["currency", "No description provided", "String"],
["deadline", "No description provided", "DateTime"],
["goal", "Goal amount in project currency", "Numeric"],
["launched", "No description provided", "DateTime"],
["pledged", "Pledged amount in the project currency", "Numeric"],
["state", "No description provided", "String"],
["backers", "No description provided", "Numeric"],
["country", "No description provided", "String"],
["usd pledged", "Pledged amount in USD (conversion made by KS)", "Numeric"],
["usd_pledged_real", "Pledged amount in USD (conversion made by fixer.io api)", "Numeric"],
["usd_goal_real", "Goal amount in USD", "Numeric"]], columns=["Field name","Field description", "Type"])

Field name                                  Field description  \
0                 ID                            No description provided   
1               name                            No description provided   
2           category                            No description provided   
3      main_category                            No description provided   
4           currency                            No description provided   
5           deadline                            No description provided   
6               goal                    Goal amount in project currency   
7           launched                            No description provided   
8            pledged             Pledged amount in the project currency   
9              state                            No description provided   
10           backers                            No description provided   
11           country                            No description provided   
12       usd pledged      Pledged amount in USD (conversion made by KS)   
13  usd_pledged_real  Pledged amount in USD (conversion made by fixe...   
14     usd_goal_real                                 Goal amount in USD   

        Type  
0    Numeric  
1     String  
2     String  
3     String  
4     String  
5   DateTime  
6    Numeric  
7   DateTime  
8    Numeric  
9     String  
10   Numeric  
11    String  
12   Numeric  
13   Numeric  
14   Numeric

* Ahora veamos los tipos de datos que detectó pandas

In [8]:
kickstarter_2018.dtypes

ID                    int64
name                 object
category             object
main_category        object
currency             object
deadline             object
goal                float64
launched             object
pledged             float64
state                object
backers               int64
country              object
usd pledged         float64
usd_pledged_real    float64
usd_goal_real       float64
dtype: object

* Los campos object generalmente son String
* Vemos que no reconoció como fechas a las variables: **deadline** y **launched** 

In [9]:
# Vemos un resumen de los datos

kickstarter_2018.describe()

ID         goal     pledged   backers  usd pledged  \
count     378661.00    378661.00   378661.00 378661.00    374864.00   
mean  1074731191.99     49080.79     9682.98    105.62      7036.73   
std    619086204.32   1183391.26    95636.01    907.19     78639.75   
min         5971.00         0.01        0.00      0.00         0.00   
25%    538263516.00      2000.00       30.00      2.00        16.98   
50%   1075275634.00      5200.00      620.00     12.00       394.72   
75%   1610148624.00     16000.00     4076.00     56.00      3034.09   
max   2147476221.00 100000000.00 20338986.27 219382.00  20338986.27   

       usd_pledged_real  usd_goal_real  
count         378661.00      378661.00  
mean            9058.92       45454.40  
std            90973.34     1152950.06  
min                0.00           0.01  
25%               31.00        2000.00  
50%              624.33        5500.00  
75%             4050.00       15500.00  
max         20338986.27   166361390.71

In [10]:
# Por defecto se ven los datos numéricos
# Pediremos ver el resto de los datos

kickstarter_2018.describe(include=['object'])

name        category main_category currency  \
count                     378657          378661        378661   378661   
unique                    375764             159            15       14   
top     New EP/Music Development  Product Design  Film & Video      USD   
freq                          41           22314         63585   295365   

          deadline             launched   state country  
count       378661               378661  378661  378661  
unique        3164               378089       6      23  
top     2014-08-08  1970-01-01 01:00:00  failed      US  
freq           705                    7  197719  292627

In [11]:
# Operamos sobre los datos pidiendo el valor mínimo

kickstarter_2018['launched'].min()

'1970-01-01 01:00:00'

In [12]:
# Operamos sobre los datos pidiendo el valor máximo
kickstarter_2018['launched'].max()

'2018-01-02 15:02:31'

In [13]:
# Indicamos que columnas corresponden a fechas

kickstarter_2018 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv", 
                               parse_dates=["deadline","launched"])
kickstarter_2018.dtypes

ID                           int64
name                        object
category                    object
main_category               object
currency                    object
deadline            datetime64[ns]
goal                       float64
launched            datetime64[ns]
pledged                    float64
state                       object
backers                      int64
country                     object
usd pledged                float64
usd_pledged_real           float64
usd_goal_real              float64
dtype: object

In [14]:
# Ahora podemos calcular el rango de la fecha de lanzamiento -> 'launched'

kickstarter_2018['launched'].max() - kickstarter_2018['launched'].min()

Timedelta('17533 days 14:02:31')

# 1.2. Asegurar de tener ids/claves únicas

* Chequeo de datos duplicados

In [15]:
kickstarter_2018.shape

(378661, 15)

* Pandas soporta índices (id) en los DataFrames

In [16]:
# Recargamos el conjunto de datos indicando que columna será el índice (index_col=['ID'])

kickstarter_2018 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv", 
                               parse_dates=["deadline","launched"],
                               index_col=['ID'])

In [17]:
kickstarter_2018[:5]

name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

           main_category currency   deadline     goal            launched  \
ID                                                                          
1000002330    Publishing      GBP 2015-10-09  1000.00 2015-08-11 12:12:28   
1000003930  Film & Video      USD 2017-11-01 30000.00 2017-09-02 04:43:57   
1000004038  Film & Video      USD 2013-02-26 45000.00 2013-01-12 00:20:50   
1000007540         Music      USD 2012-04-16  5000.00 2012-03-17 03:24:11   
1000011046  Film & Video      USD 2015-08-29 19500.00 2015-07-04 08:35:03   

            pledged     state  backers country  usd pledged  usd_pledged_real  \
ID                                                                              
1000002330     0.00    failed        0      GB         0.00              0.00   
1000003930  2421.00    failed       15      US       100.00           2421.00   
1000004038   220.00    failed        3      US       220.00            220.00   
1000007540     1.00    failed        1      US         1.00              1.00   
1000011046  1283.00  canceled       14      US      1283.00           1283.00   

            usd_goal_real  
ID                         
1000002330        1533.95  
1000003930       30000.00  
1000004038       45000.00  
1000007540        5000.00  
1000011046       19500.00

In [18]:
kickstarter_2018.shape

(378661, 14)

* Indicar el índice nos permite que podamos usarlo para buscar por ese dato

In [19]:
kickstarter_2018.loc[999988282]

name                Nou Renmen Ayiti!  We Love Haiti!
category                              Performance Art
main_category                                     Art
currency                                          USD
deadline                          2011-08-16 00:00:00
goal                                          2000.00
launched                          2011-07-19 09:07:47
pledged                                        524.00
state                                          failed
backers                                            17
country                                            US
usd pledged                                    524.00
usd_pledged_real                               524.00
usd_goal_real                                 2000.00
Name: 999988282, dtype: object

* Buscar por índice, nos permite verificar si hay filas de contenidos duplicado

In [20]:
# Para buscar duplicados

kickstarter_2018[kickstarter_2018.duplicated()]

Empty DataFrame
Columns: [name, category, main_category, currency, deadline, goal, launched, pledged, state, backers, country, usd pledged, usd_pledged_real, usd_goal_real]
Index: []

* No detectó duplicados
* Como Pandas acepta valores duplicados en los índices también debemos verificarlos

In [21]:
# Estadísticos descriptivos de la columna ID

pd.Series(kickstarter_2018.index,dtype=str).describe()

count       378661.00
mean    1074731191.99
std      619086204.32
min           5971.00
25%      538263516.00
50%     1075275634.00
75%     1610148624.00
max     2147476221.00
Name: ID, dtype: float64

In [22]:
# Buscamos nuevamente si hay datos duplicados

kickstarter_2018[kickstarter_2018.index.duplicated()]

Empty DataFrame
Columns: [name, category, main_category, currency, deadline, goal, launched, pledged, state, backers, country, usd pledged, usd_pledged_real, usd_goal_real]
Index: []

* No detectó duplicados buscando por ID

## Repasamos con un ejemplo

In [23]:
csv='1,2\n3,3\n1,3'
print(csv)

1,2
3,3
1,3


In [24]:
# pasamos los datos a una tabla
from io import StringIO
df = pd.read_csv(StringIO(csv), names=['id','value'], index_col='id')
df

value
id       
1       2
3       3
1       3

In [25]:
# Buscamos duplicados, solo muestra el primer duplicado
df[df.duplicated()]

value
id       
1       3

In [26]:
# si quremos ver todos los valores repetidos hay que usar keep=False
df[df.index.duplicated(keep=False)]

value
id       
1       2
1       3

### Ejercicio 1:

* Armar una tabla con todos los proyectos con nombres duplicados, ordenados para revisar agrupados. 

In [27]:
# Recargamos la base de datos indicando que la columna índice será el Nombre -> (index_col=['name'])

kickstarter_2018_name = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv", 
                               parse_dates=["deadline","launched"],
                               index_col=['name'])
kickstarter_2018_name[:4]

ID        category  \
name                                                                            
The Songs of Adelaide & Abullah                    1000002330          Poetry   
Greeting From Earth: ZGAC Arts Capsule For ET      1000003930  Narrative Film   
Where is Hank?                                     1000004038  Narrative Film   
ToshiCapital Rekordz Needs Help to Complete Album  1000007540           Music   

                                                  main_category currency  \
name                                                                       
The Songs of Adelaide & Abullah                      Publishing      GBP   
Greeting From Earth: ZGAC Arts Capsule For ET      Film & Video      USD   
Where is Hank?                                     Film & Video      USD   
ToshiCapital Rekordz Needs Help to Complete Album         Music      USD   

                                                    deadline     goal  \
name                                                                    
The Songs of Adelaide & Abullah                   2015-10-09  1000.00   
Greeting From Earth: ZGAC Arts Capsule For ET     2017-11-01 30000.00   
Where is Hank?                                    2013-02-26 45000.00   
ToshiCapital Rekordz Needs Help to Complete Album 2012-04-16  5000.00   

                                                             launched  \
name                                                                    
The Songs of Adelaide & Abullah                   2015-08-11 12:12:28   
Greeting From Earth: ZGAC Arts Capsule For ET     2017-09-02 04:43:57   
Where is Hank?                                    2013-01-12 00:20:50   
ToshiCapital Rekordz Needs Help to Complete Album 2012-03-17 03:24:11   

                                                   pledged   state  backers  \
name                                                                          
The Songs of Adelaide & Abullah                       0.00  failed        0   
Greeting From Earth: ZGAC Arts Capsule For ET      2421.00  failed       15   
Where is Hank?                                      220.00  failed        3   
ToshiCapital Rekordz Needs Help to Complete Album     1.00  failed        1   

                                                  country  usd pledged  \
name                                                                     
The Songs of Adelaide & Abullah                        GB         0.00   
Greeting From Earth: ZGAC Arts Capsule For ET          US       100.00   
Where is Hank?                                         US       220.00   
ToshiCapital Rekordz Needs Help to Complete Album      US         1.00   

                                                   usd_pledged_real  \
name                                                                  
The Songs of Adelaide & Abullah                                0.00   
Greeting From Earth: ZGAC Arts Capsule For ET               2421.00   
Where is Hank?                                               220.00   
ToshiCapital Rekordz Needs Help to Complete Album              1.00   

                                                   usd_goal_real  
name                                                              
The Songs of Adelaide & Abullah                          1533.95  
Greeting From Earth: ZGAC Arts Capsule For ET           30000.00  
Where is Hank?                                          45000.00  
ToshiCapital Rekordz Needs Help to Complete Album        5000.00

In [28]:
# Creamos un nuevo dataset que tenga los registros con nombres duplicados
# Buscamos nuevamente si hay datos duplicados

kickstarter_2018_dup = kickstarter_2018[kickstarter_2018.name.duplicated(keep=False)].sort_values('name')
kickstarter_2018_dup[:4]

name     category  \
ID                                                                           
816998285   "...The Last shall be first..." LODB Lifestyle...      Fashion   
815783250   "...The Last shall be first..." LODB Lifestyle...      Fashion   
1010584633                                    "A Fresh Start"       Shorts   
713417995                                     "A Fresh Start"  Documentary   

           main_category currency   deadline    goal            launched  \
ID                                                                         
816998285        Fashion      USD 2014-02-09 6500.00 2013-12-19 21:51:50   
815783250        Fashion      USD 2014-03-01 6500.00 2014-01-03 20:39:16   
1010584633  Film & Video      USD 2011-09-25 3000.00 2011-08-28 21:27:52   
713417995   Film & Video      USD 2013-03-27 5000.00 2013-01-26 03:51:47   

            pledged       state  backers country  usd pledged  \
ID                                                              
816998285      0.00    canceled        0      US         0.00   
815783250     37.00      failed        3      US        37.00   
1010584633  3000.00  successful       24      US      3000.00   
713417995   1417.00      failed       26      US      1417.00   

            usd_pledged_real  usd_goal_real  
ID                                           
816998285               0.00        6500.00  
815783250              37.00        6500.00  
1010584633           3000.00        3000.00  
713417995            1417.00        5000.00

In [29]:
kickstarter_2018_dup.shape

(5125, 14)

* Tenemos **5125** registros con nombres duplicados

# 1.3. Despersonalizar datos y guardarlos en un nuevo archivo

Hay muchas técnicas para despersonalizar datos.

Para ilustrar mostramos las ofrecidas por Google https://cloud.google.com/dlp/docs/transformations-reference:

* **Reemplazo**: Reemplaza cada valor de entrada con un valor determinado.
* **Ocultamiento**: Quita un valor y lo oculta.
* **Enmascaramiento con caracteres**: Enmascara una string por completo o parcialmente mediante el reemplazo de un número determinado de caracteres con un carácter fijo especificado..
* **Seudonimización mediante el reemplazo de un valor de entrada con un hash criptográfico**: Reemplaza valores de entrada con una string hexadecimal de 32 bytes mediante una clave de encriptación de datos.
* **Cambio de fechas**: Cambia las fechas por un número de días al azar, con la opción de ser coherente en el mismo contexto..
* **Seudonimización mediante el reemplazo con token de preservación de formato criptográfico**: Reemplaza un valor de entrada con un token, o valor sustituto, de la misma longitud mediante la encriptación de preservación de formato (FPE) con el modo de operación FFX. Esto permite que se use el resultado en sistemas con validación de formato o que necesitan aparecer como reales a pesar de que la información no se revela.
* **Valores de depósito con base en rangos de tamaño fijos**: Enmascara los valores de entrada y los reemplaza por depósitos, o rangos dentro de los cuales se encuentra el valor de entrada.
* **Valores de depósito con base en rangos de tamaño personalizados**: Valores de entrada de depósito con base en rangos configurables por el usuario y valores de reemplazo.
* **Extracción de datos de tiempo**: EExtrae o preserva una porción de los valores Date, Timestamp y TimeOfDay.

In [30]:
# Librería para hacer hashs -> códigos para encriptar nombres y despersonalizar los datos
from hashlib import md5

In [31]:
# Es necesario setear el encoding utf-8 andes de hashear

# Los floats no tienen atributo 'encode' por eso los debemos pasar a str

def hashit(val):
    if isinstance(val, float): 
        return str(val)
    return md5(val.encode('utf-8')).hexdigest()


kickstarter_2018['name'].apply(hashit)

ID
1000002330    a6828ae8a2eca25f0dd7035efc0af0a0
1000003930    81609b3bdc0b96f429672d69702f2524
1000004038    c12f5c3bace2f0213cdb2679a265dca0
1000007540    4dbdcf09c86bbf5683ec39bc57b77f81
1000011046    9c01404a2ef702811c2088ce139042ad
1000014025    3a26b565c8396d4867dda27816836e04
1000023410    1602dce9289c950176c41ce124944335
1000030581    1abb0a27e1a3ce5a79965e3022c6bddc
1000034518    3b3a87aebec0d9c1b3a4a5e4da7a88a8
100004195     d7c53ef574a108d857d52087bb685244
100004721     eba20b66506f511bfacfe843176f5e15
100005484     9df25d0c0e2f8ea9de303f3d4efaf3c8
1000055792    717f62eae71408f00fa49227da83feae
1000056157    6532f0fa53550fd516921a2d7a1121d6
1000057089    799a510f874b6d47c1fde736d22cae3e
1000064368    93ebbffc751f9cc8d31e0147b7b1c7e3
1000064918    c10592831d537538905aefc12d54b49d
1000068480    0f109f9875645b801843775bbeadb1f1
1000070642    2d3d4b4267af87948bf7ac6d3ca13b54
1000071625    26c10e7de5df6269e71df1bf49ff4da9
1000072011    459cec5845add6a7f8fb1498cfc0e618
1000081649

### Ejercicio 2:

* Verificar que los proyectos que tienen nombres duplicados también tienen el hash de nombre duplicado

In [32]:
kickstarter_2018_dup.shape

(5125, 14)

In [33]:
# Verificamos si estos registros también tienen el hash de nombre duplicado

kickstarter_2018_dup['name_hash'] = kickstarter_2018_dup['name'].apply(hashit)
kickstarter_2018_dup[:4]

name     category  \
ID                                                                           
816998285   "...The Last shall be first..." LODB Lifestyle...      Fashion   
815783250   "...The Last shall be first..." LODB Lifestyle...      Fashion   
1010584633                                    "A Fresh Start"       Shorts   
713417995                                     "A Fresh Start"  Documentary   

           main_category currency   deadline    goal            launched  \
ID                                                                         
816998285        Fashion      USD 2014-02-09 6500.00 2013-12-19 21:51:50   
815783250        Fashion      USD 2014-03-01 6500.00 2014-01-03 20:39:16   
1010584633  Film & Video      USD 2011-09-25 3000.00 2011-08-28 21:27:52   
713417995   Film & Video      USD 2013-03-27 5000.00 2013-01-26 03:51:47   

            pledged       state  backers country  usd pledged  \
ID                                                              
816998285      0.00    canceled        0      US         0.00   
815783250     37.00      failed        3      US        37.00   
1010584633  3000.00  successful       24      US      3000.00   
713417995   1417.00      failed       26      US      1417.00   

            usd_pledged_real  usd_goal_real                         name_hash  
ID                                                                             
816998285               0.00        6500.00  0c7a251ffe4c7834cbc4b04906952ff1  
815783250              37.00        6500.00  0c7a251ffe4c7834cbc4b04906952ff1  
1010584633           3000.00        3000.00  67554ab4203d95f2f2f05365f768206e  
713417995            1417.00        5000.00  67554ab4203d95f2f2f05365f768206e

In [34]:
kickstarter_2018_dup.shape

(5125, 15)

* Tenemos **5125** registros con hash duplicados
* Estos registros se corresponden con los que habíamos detectado para nombres duplicados

# 1.4. Nunca modificar los datos crudos u originales


In [35]:
#no hacer modificaciones para no perder los datos originales

#if not in_colab:
#  kickstarter_2018.to_csv(BASE_DIR + "/output/ks-projects-201801-for-pandas.csv")


# 2. Pasos necesarios


In [36]:
# Paso necesario si no estamos en colab

import pandas as pd

if not in_colab:
  kickstarter_2018 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801-for-pandas.csv",
  #kickstarter_2018 = pd.read_csv(BASE_DIR + "/output/ks-projects-201801-for-pandas.csv",
                        index_col='ID',
                        parse_dates=['deadline','launched'])

In [37]:
kickstarter_2018.describe(include='all')

name        category main_category currency  \
count                     378657          378661        378661   378661   
unique                    375764             159            15       14   
top     New EP/Music Development  Product Design  Film & Video      USD   
freq                          41           22314         63585   295365   
first                        NaN             NaN           NaN      NaN   
last                         NaN             NaN           NaN      NaN   
mean                         NaN             NaN           NaN      NaN   
std                          NaN             NaN           NaN      NaN   
min                          NaN             NaN           NaN      NaN   
25%                          NaN             NaN           NaN      NaN   
50%                          NaN             NaN           NaN      NaN   
75%                          NaN             NaN           NaN      NaN   
max                          NaN             NaN           NaN      NaN   

                   deadline         goal             launched     pledged  \
count                378661    378661.00               378661   378661.00   
unique                 3164          nan               378089         nan   
top     2014-08-08 00:00:00          nan  1970-01-01 01:00:00         nan   
freq                    705          nan                    7         nan   
first   2009-05-03 00:00:00          nan  1970-01-01 01:00:00         nan   
last    2018-03-03 00:00:00          nan  2018-01-02 15:02:31         nan   
mean                    NaN     49080.79                  NaN     9682.98   
std                     NaN   1183391.26                  NaN    95636.01   
min                     NaN         0.01                  NaN        0.00   
25%                     NaN      2000.00                  NaN       30.00   
50%                     NaN      5200.00                  NaN      620.00   
75%                     NaN     16000.00                  NaN     4076.00   
max                     NaN 100000000.00                  NaN 20338986.27   

         state   backers country  usd pledged  usd_pledged_real  usd_goal_real  
count   378661 378661.00  378661    374864.00         378661.00      378661.00  
unique       6       nan      23          nan               nan            nan  
top     failed       nan      US          nan               nan            nan  
freq    197719       nan  292627          nan               nan            nan  
first      NaN       nan     NaN          nan               nan            nan  
last       NaN       nan     NaN          nan               nan            nan  
mean       NaN    105.62     NaN      7036.73           9058.92       45454.40  
std        NaN    907.19     NaN     78639.75          90973.34     1152950.06  
min        NaN      0.00     NaN         0.00              0.00           0.01  
25%        NaN      2.00     NaN        16.98             31.00        2000.00  
50%        NaN     12.00     NaN       394.72            624.33        5500.00  
75%        NaN     56.00     NaN      3034.09           4050.00       15500.00  
max        NaN 219382.00     NaN  20338986.27       20338986.27   166361390.71

## 2.1. En las etiquetas de las variables, no se deben usar caracteres especiales.

In [38]:
# helpful character encoding module
import chardet

In [39]:
# Los caracteres 'normales' son los números y las letras
import string

string.ascii_letters + string.digits 

'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'

* ASCII: Es una de las primeras codificaciones (o encoding) disponibles. Ver https://en.wikipedia.org/wiki/ASCII

* Un encoding es un mapa de caracteres a una representación en bits (por ejemplo 1000001). ASCII es uno de los primeros estandares para interoperatividad entre computadoras (antes cada fabricante de computadora usa su propia codificación), contempla sólo los caracteres ingleses y usa 7 bits, por ejemplo *1000001* codifica el caracter *A*

* Cómo sólo representaban los caracteres ingleses, empiezan a aparecer variantes para distintos idiomas, Microsoft hizo los propios para internacionalizar Windows, hubo otro estandar ISO 8859, hasta que llegó el éstandar Unicode que hizo múltiples encodings pero cada uno contempla múltiples idiomas e incluso es extensible a futuro. UTF-8 es de esos encodings el más utilizado. Ver https://en.wikipedia.org/wiki/Unicode

* En particular, Python 3 utiliza UTF-8 por defecto (lo que no sucedía en Python 2).

### Veamos un ejemplo práctico

In [40]:
# Comenzar por un string
before = "This is the euro symbol: €"

# checkear que datatype es
type(before)

str

In [41]:
# cambiar el encode reemplazando caracteres que den errores
after = before.encode("utf-8", errors = "replace")

# checkear que datatype es
type(after)

bytes

In [42]:
# veamos como se ven los bytes
after

# cuando no tiene un simbolo ascii python muestra el codigo exadecimal del símbolo (\xe2\x82\xac)

b'This is the euro symbol: \xe2\x82\xac'

In [43]:
# reconvertirlo a utf-8
print(after.decode("utf-8"))

This is the euro symbol: €


In [44]:
# decodifica los bytes con el ascii encoding

#print(after.decode("ascii"))
after_ascii = after.decode("ascii", errors = "replace")

In [45]:
# start with a string
before = "This is the euro symbol: €"

# encode it to a different encoding, replacing characters that raise errors
after = before.encode("utf-8", errors = "replace")

# convert it back to utf-8
print(after.decode("utf-8"))

# We've lost the original underlying byte string! It's been 
# replaced with the underlying byte string for the unknown character :(

This is the euro symbol: €


* Cuando obtenemos un error `UnicodeDecodeError` nos indica que el archivo no estaría en Unicode. 
* Si no está en algún metadato o no fue indicado es dificil detectar el encoding, para ello utilizamos un paquete de Python, que aplica unas heurísticas para detectar encoding.

In [46]:
# Módulo para detectar el encoding

import chardet

if in_colab:
  !wget -c $BASE_DIR/input/kickstarter-projects/ks-projects-201612.csv
  DOWN_DIR = "./"
else:
  DOWN_DIR = BASE_DIR + "/input/kickstarter-projects/"
  
with open(DOWN_DIR + "/ks-projects-201612.csv", 'rb') as rawdata:
    result = chardet.detect(rawdata.read(10000))

result

{'encoding': 'Windows-1252', 'confidence': 0.73, 'language': ''}

* El encoding que posee la base de datos es: **Windows-1252**
* Entonces, una vez detectado, le podemos indicar el encoding al leer el archivos

In [47]:
# Leemos la base de datos con el encoding detectado por el módulo chardet

kickstarter_2016 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201612.csv", 
                               encoding='Windows-1252')
kickstarter_2016.head()

C:\Users\W10\miniconda3\envs\diplodatos-ayvd\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (13,14,15) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


ID                                               name   \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000004038                                     Where is Hank?   
2  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
3  1000011046  Community Film Project: The Art of Neighborhoo...   
4  1000014025                               Monarch Espresso Bar   

        category  main_category  currency             deadline   goal   \
0          Poetry     Publishing       GBP  2015-10-09 11:36:00   1000   
1  Narrative Film   Film & Video       USD  2013-02-26 00:20:50  45000   
2           Music          Music       USD  2012-04-16 04:24:11   5000   
3    Film & Video   Film & Video       USD  2015-08-29 01:00:00  19500   
4     Restaurants           Food       USD  2016-04-01 13:38:27  50000   

             launched  pledged       state  backers  country  usd pledged   \
0  2015-08-11 12:12:28        0      failed        0       GB            0   
1  2013-01-12 00:20:50      220      failed        3       US          220   
2  2012-03-17 03:24:11        1      failed        1       US            1   
3  2015-07-04 08:35:03     1283    canceled       14       US         1283   
4  2016-02-26 13:38:27    52375  successful      224       US        52375   

  Unnamed: 13 Unnamed: 14 Unnamed: 15  Unnamed: 16  
0         NaN         NaN         NaN          nan  
1         NaN         NaN         NaN          nan  
2         NaN         NaN         NaN          nan  
3         NaN         NaN         NaN          nan  
4         NaN         NaN         NaN          nan

* Otro problema con los caracteres que es más sutíl es el: **[Mojibake](https://en.wikipedia.org/wiki/Mojibake)**

* Sucede al leer UTF-8 como ASCII extendido (ie. `Windows-*`, `iso-8859-*` entre otros)

### Ejemplo de mojibake:

In [48]:
"cigüeña".encode('utf-8').decode('iso-8859-1')

'cigüeña'

* Para solucionar esto vamos a usar un módulo específico: [ftfy](https://ftfy.readthedocs.io/en/latest/). 

* Este módulo nos permite ordenar cadenas por rareza

In [49]:
import ftfy.badness as bad

In [50]:
# Buscamos valores raros en la variable 'name' de la base de datos

def weird(val):
    if isinstance(val, float): 
        return 0
    return bad.sequence_weirdness(val)


kickstarter_2018['name_weirdness'] = kickstarter_2018['name'].apply(weird)
kickstarter_2018[kickstarter_2018['name_weirdness'] > 1]

name  \
ID                                                              
1221990593             Support PMCD´s ¨From a Different Clay¨   
1234089244  Cracking (Mastering) the Metaphor 
(Political)...   
124795481                                    µGALE (Canceled)   
1468489999  µCurrent GOLD - Precision Multimeter Current A...   
1610503461  µPeek - The Professional Microscope that Fits ...   
16174550                                       PARALELO 19 Nº   
1719215256           Kickstarter Gold: Manual Coffeemaker Nº3   
1837094178  ¨Midgard's taigh-seinnse/ Eröffnung Mittelalte...   
2056123754      ZEIº - the most simple time tracking solution   
2068157547  Ballet Folklorico Infantil de la Universidad ...   
2096648634                          Funky City : a funky rpg   
2102265118      Music Album ¨Dreams¨ Uplifting Tunes with Rap   
237780030   Solderdoodle Pro: Over 700ºF USB Rechargeable ...   
383244441            Virtual Roller Coaster in 4th Dimension   
458626388   Solderdoodle Pro: 700ºF USB Rechargeable Solde...   
466898065                                Manual Tea Maker Nº1   
516833776    µTracker - extremly Power Saving GPS + GSM Board   
549309727   Higher Grounds Coffee Company Mobile Coffee Tr...   
551994337   ?? cybertwee ?? ¨*:·. bake sale on the deep we...   
558638957                                         Mµ Magazine   
718763713      Not Twins - New EP! "The View from Down Here"   
805839986   µRuler - Measure size in the micro-world under...   
900880068                                Newton ¨El Veterano¨   
960506632                              Degreesº to Extinction   
963442008          theNewerYork Lit Mag: We M?§§ w/ Yoµr H£åd   
98169605    µTøøl - The Ultimate pocket tool! Credit card ...   

                   category main_category currency   deadline      goal  \
ID                                                                        
1221990593            Dance         Dance      USD 2011-06-08   1500.00   
1234089244       Nonfiction    Publishing      USD 2013-10-27  30000.00   
124795481         Wearables    Technology      EUR 2015-04-17  18000.00   
1468489999         Hardware    Technology      AUD 2013-12-31   9900.00   
1610503461       Technology    Technology      CHF 2016-01-09 125000.00   
16174550        Photography   Photography      MXN 2017-04-15  15000.00   
1719215256   Product Design        Design      USD 2017-07-22  60000.00   
1837094178      Restaurants          Food      CHF 2016-09-29  50000.00   
2056123754       Technology    Technology      EUR 2016-10-26  75000.00   
2068157547              Art           Art      MXN 2017-04-03  40000.00   
2096648634      Video Games         Games      USD 2017-04-08   5000.00   
2102265118          Hip-Hop         Music      EUR 2017-12-24  28000.00   
237780030   DIY Electronics    Technology      USD 2014-08-19  37800.00   
383244441       Makerspaces    Technology      MXN 2017-02-25 459900.00   
458626388   DIY Electronics    Technology      USD 2014-09-24  11400.00   
466898065    Product Design        Design      USD 2016-08-20  40000.00   
516833776          Hardware    Technology      EUR 2016-12-09  89000.00   
549309727       Food Trucks          Food      USD 2018-01-12  10000.00   
551994337       Digital Art           Art      USD 2015-09-18    100.00   
558638957             Zines    Publishing      CAD 2016-12-02   5000.00   
718763713        Indie Rock         Music      USD 2017-03-27   3500.00   
805839986        Technology    Technology      CAD 2017-04-07   2000.00   
900880068             Music         Music      EUR 2016-12-10   2747.00   
960506632       Documentary  Film & Video      USD 2016-12-21  30000.00   
963442008       Periodicals    Publishing      USD 2012-03-25   7000.00   
98169605         Technology    Technology      NZD 2016-01-01  20000.00   

                      launched   pledged       state  backers country  \
ID                                                            

* Tenemos **26 valores raros** en la variable 'name' de la base de datos

In [51]:
# Chequeamos que no haya caracteres fuera de a-Z, 0-9 y _ en los nombres de las variables/columnas

kickstarter_2018.columns[~kickstarter_2018.columns.str.match(r'^(\w+)$')]

Index(['usd pledged'], dtype='object')

* Tenemos una columna en cuyo nombre hay un espacio vacío(usd pledged), 
* Este espacio en el nombre de la variable, no nos causará inconvenientes. por eso asignaremos un guión bajo "_"  donde hay un espacio vacío (usd_pledged).

In [52]:
# Reemplazamos el espacio vacío por un guión bajo

kickstarter_2018.columns = kickstarter_2018.columns.str.replace(' ', '_')
kickstarter_2018.head()

name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

           main_category currency   deadline     goal            launched  \
ID                                                                          
1000002330    Publishing      GBP 2015-10-09  1000.00 2015-08-11 12:12:28   
1000003930  Film & Video      USD 2017-11-01 30000.00 2017-09-02 04:43:57   
1000004038  Film & Video      USD 2013-02-26 45000.00 2013-01-12 00:20:50   
1000007540         Music      USD 2012-04-16  5000.00 2012-03-17 03:24:11   
1000011046  Film & Video      USD 2015-08-29 19500.00 2015-07-04 08:35:03   

            pledged     state  backers country  usd_pledged  usd_pledged_real  \
ID                                                                              
1000002330     0.00    failed        0      GB         0.00              0.00   
1000003930  2421.00    failed       15      US       100.00           2421.00   
1000004038   220.00    failed        3      US       220.00            220.00   
1000007540     1.00    failed        1      US         1.00              1.00   
1000011046  1283.00  canceled       14      US      1283.00           1283.00   

            usd_goal_real  name_weirdness  
ID                                         
1000002330        1533.95               0  
1000003930       30000.00               0  
1000004038       45000.00               0  
1000007540        5000.00               0  
1000011046       19500.00               0

In [53]:
kickstarter_2018.usd_pledged.head()

ID
1000002330      0.00
1000003930    100.00
1000004038    220.00
1000007540      1.00
1000011046   1283.00
Name: usd_pledged, dtype: float64

### Ejercicio 3

* Comparar la cantidad de nombres raros en kickstarter_2018 con la que obtenemos al cargar 'ks-projects-201801.csv' con encoding iso-8859-1.

In [54]:
# Importamos la base de datos con encoding iso-8859-1

# kickstarter_2018 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv")

kickstarter_2018_iso_8859_1 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv", 
                                          encoding='iso-8859-1')

kickstarter_2018_iso_8859_1.columns[~kickstarter_2018_iso_8859_1.columns.str.match(r'^(\w+)$')]

Index(['usd pledged'], dtype='object')

In [55]:
# Buscamos valores raros en la columna 'name' de la base de datos con encoding iso-8859-1

def weird(val):
    if isinstance(val, float):
        return 0
    return bad.sequence_weirdness(val)

kickstarter_2018_iso_8859_1['name_weirdness'] = kickstarter_2018_iso_8859_1['name'].apply(weird)
kickstarter_2018_iso_8859_1[kickstarter_2018_iso_8859_1['name_weirdness'] > 1]

ID                                               name  \
63      1000235643                                     HIIT Bottle™   
70      1000260691  Gizbee™ Unlimited Removable Storage for Your...   
123     1000519001             Wordwright: re·de·fin·ing word play   
157     1000644119                                   Operación Douve   
184     1000767040  Strait Jacket – Tangle Prevention For Headph...   
238     1001042227                            Via Chic–OneClutch™   
325     1001484182                       Daweyu Hills – Crop to Cup   
326     1001490622                             TimeChat™ (Canceled)   
437     1002098978                             Lux3 – Wellness Lamp   
470     1002234028                 æmp:dance / amiti perry + company   
472     1002250421  Spiele für iOS und Android / Games for Androi...   
477     1002287054  Michale Graves “Vagabond Acoustic” Extreme...   
526     1002522073  Lampenschirme und Objekte aus französischer H...   
627      100301082                           Feliz año nuevo, mamá.   
646     1003127225  AcneDust™ Organic Acne Treatment -Acne Dust ...   
664     1003240719                       Boutique Disco Flash Québec   
703     1003456090                 Academia de Música impulsarte.org   
733     1003538517                      Pollock l’acchiappa-panzane   
739     1003561611                                Vyu360™ 4K ProCam   
846      100413062                               Corey’s Coffee Bar   
918     1004598691  Jakten på makten - Brädspelet om svensk poli...   
964     1004866875                       Original paintings from £10   
1024    1005175741                        UNCLE SAMta – FINAL HOURS   
1136     100576591         Nina Jo Smith’s PEOPLE PLACES & SINGS CD   
1165    1005925256  The GoShelf™ - corner shelf self installatio...   
1184    1005999335   wecemat – The ecologic automated flush adapter   
1343    1006885357  Escape from Alcatraz — A New Mystery Rummy Game   
1368    1007021695                  Broschüre Maultiermuseum Schweiz   
1374    1007043144               FLEXXBALL™ - The Playful Sculpture   
1404    1007170928  A través del arte " Identidad y Bíodiversida...   
...            ...                                                ...   
377458   994028313  Fautus • We are the UBER of photography –w...   
377525   994394262                         The Handle Wonder Cover™   
377648   994971049  Brookdale Community College Baseball Fundraise...   
377714   995318657                 Stötta Vandringsbloggen Community   
377720    99535653  BITE-SIZED CHINESE – CHARACTER COLORING BOOK...   
377834   995932608  Solenoid™ micro-compact, high fidelity speak...   
377867   996085230  SQUAREMINO™ - The Square Domino Tabletop Boa...   
377920   996430564  Best Leggings Ever! — Unique Fabric  / Styli...   
377935   996498361    All Season Sweats™ - A New Type of Sweat Pant   
377936   996506334  Clean Apparel – The Most Sustainable T-Shirt...   
377939   996519443  HIDEBOUND—world premiere puppet performances...   
378050   997083412  Train: The Early Years. A photographer’s ins...   
378053   997102132                                        BÜD Brand.   
378095   997340463                  QRawr – Comments for real walls   
378096   997343329  Drakonheim: City of Bones — A Systemless RPG...   
378113   997425910        Zulu land’s Rhino Warriors TV docu-series   
378126   997481500  Sound Cover™: iPad2 Cover with Powerful Buil...   
378175   997720090                 high fashion royal hubb bébé air   
378285    99825633          Battle Systems™ Fantasy Dungeon Terrain   
378317   998378475  [title of show] — The Chicago Storefront Pre...   
378322   998401361  Support Tiffany Monìque's New Video & Single!...   
378329   998418451                                    Miami en vidéo   
378490   999133026         Kuña: documental sobre la mujer paraguaya   
3

* Tenemos **26 valores raros** en la base de datos sin setear el encoding
* Tenemos **7682 valores raros** en la base de datos con encoding iso_8859_1

## 2.2. Tratar valores faltantes

* Veremos cuantos valores nulos tenemos

In [56]:
# Obtenemos el número de valores faltantes por columna 

missing_values_count = kickstarter_2018.isnull().sum()

# look at the # of missing points in the first ten columns

missing_values_count[missing_values_count > 0]

name              4
usd_pledged    3797
dtype: int64

* Tnemos 4 filas que no tienen nombre
* Tenemos 3797 filas sin valor en la columna usd_pledged

In [57]:
# Calculamos el impacto en términos relativos que tienen los valores nulos de nuesta base de datos

len(kickstarter_2018.dropna())/len(kickstarter_2018)

0.9899619976707398

In [58]:
# Calculamos en impacto en términos relativos que tienen los valores nulos de la columna 'name'

len(kickstarter_2018.dropna(subset=['name']))/len(kickstarter_2018)

0.9999894364616372

**Tener en cuenta:**   
* Descartar datos siempre es la ultima opcion.
* Si los datos nulos son muy pocos y consideramos que no afectan en el análisis, en ese caso los podemos quitar.

In [59]:
# Calculamos que significan en términos de representatividad e importancia de nuestro análisis, 
# esos valores nulos

kickstarter_2018[kickstarter_2018.name.isnull()]

name        category main_category currency   deadline      goal  \
ID                                                                            
1848699072  NaN  Narrative Film  Film & Video      USD 2012-02-29 200000.00   
634871725   NaN     Video Games         Games      GBP 2013-01-06   2000.00   
648853978   NaN  Product Design        Design      USD 2016-07-18   2500.00   
796533179   NaN        Painting           Art      USD 2011-12-05  35000.00   

                      launched  pledged      state  backers country  \
ID                                                                    
1848699072 2012-01-01 12:35:31   100.00     failed        1      US   
634871725  2012-12-19 23:57:48   196.00     failed       12      GB   
648853978  2016-06-18 05:01:47     0.00  suspended        0      US   
796533179  2011-11-06 23:55:55   220.00     failed        5      US   

            usd_pledged  usd_pledged_real  usd_goal_real  name_weirdness  
ID                                                                        
1848699072       100.00            100.00      200000.00               0  
634871725        317.73            316.05        3224.97               0  
648853978          0.00              0.00        2500.00               0  
796533179        220.00            220.00       35000.00               0

* **Ejemplo**, cuantos de los proyectos suspendidos no tienen nombre

In [60]:
len(kickstarter_2018[kickstarter_2018.state == 'suspended'].dropna(subset=['name'])) / \
    len(kickstarter_2018[kickstarter_2018.state == 'suspended'])

0.9994582881906826

* Solamente tenemos 4 registros sin nombre.
* En este caso, creemos que es seguro descartar los proyectos sin nombre ya que son muy pocos.

In [61]:
# Limpiamos de la base de datos los 4 resgistros con valores nulos en la variable 'name'

kickstarter_2018 = kickstarter_2018.dropna(subset=['name'])

In [62]:
# Ahora estudiaremos los datos nulos de la columna usd_pledged (3797)

kickstarter_2018[kickstarter_2018.usd_pledged.isnull()]

name      category  \
ID                                                                            
1000694855                      STREETFIGHTERZ WHEELIE MURICA  Film & Video   
100149523                         Duncan Woods - Chameleon EP         Music   
1003023003          The Making of Ashley Kelley's Debut Album         Music   
1003130892                       Butter Side Down Debut Album         Music   
1003629045                            Chase Goehring debut EP         Music   
1004013077           Spencer Capier Instrumental Project 2015         Music   
1004126342  LUKAS LIGETI'S 50TH BIRTHDAY FESTIVAL: ORIGINA...         Music   
1004255433      The Battle For Breukelen: A Neighborhood Epic  Film & Video   
1004298993  "Tamboura Plays Violin" - a collection of Pop ...         Music   
1004402863        Nightingale Noel - An A Cappella Holiday CD         Music   
1005185256                             Local Music Connection         Music   
100522240   DO NOT DUPLICATE - Selected by 2015 Devon Film...  Film & Video   
1005653464                                       SA-4 Studios         Music   
1006327667                           Letters to the Wild - EP         Music   
1008671527                          The Fitness Mindset   :-)    Publishing   
1008675685               Inspired & the Sleep EP - Eyelid Kid         Music   
1008815806  Into Winter: An Instrumental Christmas Experiment         Music   
1009154868       Help advance scientific guitar & bass design         Music   
1009312309                                        Mr. Pickles    Publishing   
1009774693                Universal Peace | Universal Justice         Music   
1010217550                               Solarboot-Havel-Tour         Music   
1010694326                                          Surrender         Music   
1011508551                               EBENEZER Solo Record         Music   
1011952280                         NHSOB Summer Festival 2015         Music   
1012058337                                        Degobah Max         Music   
1012414277    WE ARE WHO WE ARE BUT WHATEVER WE WANT TO BE!!!         Music   
1012704793                               Life In The Bus Lane  Film & Video   
1012744036          An Oratorio for our Time - Last Stop Cafe         Music   
1012807897                                  Juliet Remembered  Film & Video   
1013389538  Make admission to the London Music Awards cere...         Music   
...                                                       ...           ...   
982862687                                     Bydlo - A Novel    Publishing   
982919439                                               Faith  Film & Video   
983728691                    The Octopodes' 5th Studio Album!         Music   
985231591                                            TWO FISH  Film & Video   
986253191   DEVELOP MAGAZINE FOR THE ARTS, MUSIC, AND DESI...    Publishing   
986471034      Help Revival finish their new album, Guidance.         Music   
98744618                           Vinyl for Dancing Pigeons!         Music   
988763792     Maud Sings Maud: A Musical Betsy-Tacy Companion         Music   
989218116                    Keep on Movin' - The Debut Album         Music   
989499289                                           Sojourner         Music   
989874415                         Love and Soul through Sound         Music   
991409513                           Chris Copeland's First EP         Music   
991483230   Support Vocal Synergy's Second Semester and Te...         Music   
991820324                                   The Blend CD 2015         Music   
992433603                                   Jonny Come Lately  Film & Video   
994080962                                     Maximum Coppage  Film & Video   
994621918                                         Star Sailor  Film & Video   
994911312                     "Songwriter assistance" service         Music   
994933706                   

In [63]:
# pedimos los estadísticos descriptivos de esta variable (usd_pledged)

kickstarter_2018[kickstarter_2018.usd_pledged.isnull()].describe(include='all')

name category main_category currency  \
count                       3797     3797          3797     3797   
unique                      3764       28            12       10   
top     New EP/Music Development    Music         Music      USD   
freq                          28     2387          2388     2738   
first                        NaN      NaN           NaN      NaN   
last                         NaN      NaN           NaN      NaN   
mean                         NaN      NaN           NaN      NaN   
std                          NaN      NaN           NaN      NaN   
min                          NaN      NaN           NaN      NaN   
25%                          NaN      NaN           NaN      NaN   
50%                          NaN      NaN           NaN      NaN   
75%                          NaN      NaN           NaN      NaN   
max                          NaN      NaN           NaN      NaN   

                   deadline       goal             launched    pledged  \
count                  3797    3797.00                 3797    3797.00   
unique                  722        nan                 3797        nan   
top     2015-05-01 00:00:00        nan  2015-10-27 17:00:11        nan   
freq                     26        nan                    1        nan   
first   2012-04-24 00:00:00        nan  2012-03-14 19:05:44        nan   
last    2016-07-07 00:00:00        nan  2016-05-21 20:45:12        nan   
mean                    NaN    5413.35                  NaN    3023.26   
std                     NaN   37466.24                  NaN   44919.41   
min                     NaN     108.00                  NaN       0.00   
25%                     NaN    2000.00                  NaN       5.00   
50%                     NaN    3500.00                  NaN     640.00   
75%                     NaN    5000.00                  NaN    3845.00   
max                     NaN 1877000.00                  NaN 2762571.78   

            state  backers country  usd_pledged  usd_pledged_real  \
count        3797  3797.00    3797         0.00           3797.00   
unique          6      nan       1          nan               nan   
top     undefined      nan    N,0"          nan               nan   
freq         3562      nan    3797          nan               nan   
first         NaN      nan     NaN          nan               nan   
last          NaN      nan     NaN          nan               nan   
mean          NaN     0.00     NaN          nan           2949.85   
std           NaN     0.00     NaN          nan          44881.72   
min           NaN     0.00     NaN          nan              0.00   
25%           NaN     0.00     NaN          nan              5.00   
50%           NaN     0.00     NaN          nan            616.00   
75%           NaN     0.00     NaN          nan           3932.23   
max           NaN     0.00     NaN          nan        2762571.78   

        usd_goal_real  name_weirdness  
count         3797.00         3797.00  
unique            nan             nan  
top               nan             nan  
freq              nan             nan  
first             nan             nan  
last              nan             nan  
mean          5097.31           -0.01  
std          37258.79            0.12  
min             85.28           -3.00  
25%           2250.00            0.00  
50%           3583.05            0.00  
75%           5208.09            0.00  
max        1877000.00            0.00

**Interesante**,    
* Todos los datos pertenecen a un país indeterminado.
* Todos los datos de ese "país" no tienen dinero prometido.

In [64]:
kickstarter_2018[(kickstarter_2018.country == 'N,0"') & ~(kickstarter_2018.usd_pledged.isnull())].head()

Empty DataFrame
Columns: [name, category, main_category, currency, deadline, goal, launched, pledged, state, backers, country, usd_pledged, usd_pledged_real, usd_goal_real, name_weirdness]
Index: []

* A priori como los proyectos no tienen patrocinantes, completamos con 0 los valores faltantes

In [65]:
# Completamos con 0 los valores faltantes

kickstarter_2018 = kickstarter_2018.fillna(0)
missing_values_count = kickstarter_2018.isnull().sum()
missing_values_count[missing_values_count > 0]

Series([], dtype: int64)

In [66]:
# Vemos con cuantos datos finales nos quedamos en el dataset luego del filtrado (378657)

kickstarter_2018.shape

(378657, 15)

* **Más métodos disponibles en:** https://pandas.pydata.org/pandas-docs/stable/api.html#api-dataframe-missing y http://scikit-learn.org/stable/modules/classes.html#module-sklearn.preprocessing

## 2.3. Codificar variables

* Para trabajar con los algoritmos de aprendizaje automático, las **variables categóricas**  deben ser codificadas como **variables numéricas**, no como cadenas.

* Para esta tarea también hay diferentes estrategias, dos comunes son: 
* * asociar cadena a número,
* * asociar cadena a columna.

In [67]:
import sklearn
from sklearn import preprocessing

In [68]:
# Pedimos una descripción del tipo de variables

kickstarter_2018.describe(include=['O'])

name        category main_category currency  \
count                     378657          378657        378657   378657   
unique                    375764             159            15       14   
top     New EP/Music Development  Product Design  Film & Video      USD   
freq                          41           22313         63584   295362   

         state country  
count   378657  378657  
unique       6      23  
top     failed      US  
freq    197716  292624

In [69]:
column='main_category'

# Create a label (category) encoder object
le = preprocessing.LabelEncoder()

# Fit the encoder to the pandas column
le.fit(kickstarter_2018[column])

LabelEncoder()

In [70]:
# View encoder mapping
dict(zip(le.classes_,le.transform(le.classes_)))

{'Art': 0,
 'Comics': 1,
 'Crafts': 2,
 'Dance': 3,
 'Design': 4,
 'Fashion': 5,
 'Film & Video': 6,
 'Food': 7,
 'Games': 8,
 'Journalism': 9,
 'Music': 10,
 'Photography': 11,
 'Publishing': 12,
 'Technology': 13,
 'Theater': 14}

In [71]:
# Apply the fitted encoder to the pandas column
kickstarter_2018[column] = le.transform(kickstarter_2018[column]) 
kickstarter_2018.head()

name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

            main_category currency   deadline     goal            launched  \
ID                                                                           
1000002330             12      GBP 2015-10-09  1000.00 2015-08-11 12:12:28   
1000003930              6      USD 2017-11-01 30000.00 2017-09-02 04:43:57   
1000004038              6      USD 2013-02-26 45000.00 2013-01-12 00:20:50   
1000007540             10      USD 2012-04-16  5000.00 2012-03-17 03:24:11   
1000011046              6      USD 2015-08-29 19500.00 2015-07-04 08:35:03   

            pledged     state  backers country  usd_pledged  usd_pledged_real  \
ID                                                                              
1000002330     0.00    failed        0      GB         0.00              0.00   
1000003930  2421.00    failed       15      US       100.00           2421.00   
1000004038   220.00    failed        3      US       220.00            220.00   
1000007540     1.00    failed        1      US         1.00              1.00   
1000011046  1283.00  canceled       14      US      1283.00           1283.00   

            usd_goal_real  name_weirdness  
ID                                         
1000002330        1533.95               0  
1000003930       30000.00               0  
1000004038       45000.00               0  
1000007540        5000.00               0  
1000011046       19500.00               0

In [72]:
# Reversing encoding
kickstarter_2018[column] = le.inverse_transform(kickstarter_2018[column]) 
kickstarter_2018.head()

name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

           main_category currency   deadline     goal            launched  \
ID                                                                          
1000002330    Publishing      GBP 2015-10-09  1000.00 2015-08-11 12:12:28   
1000003930  Film & Video      USD 2017-11-01 30000.00 2017-09-02 04:43:57   
1000004038  Film & Video      USD 2013-02-26 45000.00 2013-01-12 00:20:50   
1000007540         Music      USD 2012-04-16  5000.00 2012-03-17 03:24:11   
1000011046  Film & Video      USD 2015-08-29 19500.00 2015-07-04 08:35:03   

            pledged     state  backers country  usd_pledged  usd_pledged_real  \
ID                                                                              
1000002330     0.00    failed        0      GB         0.00              0.00   
1000003930  2421.00    failed       15      US       100.00           2421.00   
1000004038   220.00    failed        3      US       220.00            220.00   
1000007540     1.00    failed        1      US         1.00              1.00   
1000011046  1283.00  canceled       14      US      1283.00           1283.00   

            usd_goal_real  name_weirdness  
ID                                         
1000002330        1533.95               0  
1000003930       30000.00               0  
1000004038       45000.00               0  
1000007540        5000.00               0  
1000011046       19500.00               0

In [73]:
from sklearn.preprocessing import LabelBinarizer
other_column = 'state'

lb = LabelBinarizer()

lb_results = lb.fit_transform(kickstarter_2018[other_column])
pd.DataFrame(lb_results, columns=((other_column + '_') + pd.Series(lb.classes_))).head(10)

state_canceled  state_failed  state_live  state_successful  \
0               0             1           0                 0   
1               0             1           0                 0   
2               0             1           0                 0   
3               0             1           0                 0   
4               1             0           0                 0   
5               0             0           0                 1   
6               0             0           0                 1   
7               0             1           0                 0   
8               1             0           0                 0   
9               1             0           0                 0   

   state_suspended  state_undefined  
0                0                0  
1                0                0  
2                0                0  
3                0                0  
4                0                0  
5                0                0  
6                0                0  
7                0                0  
8                0                0  
9                0                0

* More about **preprocessing** in: http://scikit-learn.org/stable/modules/classes.html#module-sklearn.preprocessing

#### Ejercicio 4

Codificar `currency` con ambas estrategias

## 2.4. No cambiar los nombres de las variables de la fuente de origen


## 2.5. Verificar la consistencia de las variables
Aplicar reglas de integridad


In [74]:
kickstarter_2018.head()

name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

           main_category currency   deadline     goal            launched  \
ID                                                                          
1000002330    Publishing      GBP 2015-10-09  1000.00 2015-08-11 12:12:28   
1000003930  Film & Video      USD 2017-11-01 30000.00 2017-09-02 04:43:57   
1000004038  Film & Video      USD 2013-02-26 45000.00 2013-01-12 00:20:50   
1000007540         Music      USD 2012-04-16  5000.00 2012-03-17 03:24:11   
1000011046  Film & Video      USD 2015-08-29 19500.00 2015-07-04 08:35:03   

            pledged     state  backers country  usd_pledged  usd_pledged_real  \
ID                                                                              
1000002330     0.00    failed        0      GB         0.00              0.00   
1000003930  2421.00    failed       15      US       100.00           2421.00   
1000004038   220.00    failed        3      US       220.00            220.00   
1000007540     1.00    failed        1      US         1.00              1.00   
1000011046  1283.00  canceled       14      US      1283.00           1283.00   

            usd_goal_real  name_weirdness  
ID                                         
1000002330        1533.95               0  
1000003930       30000.00               0  
1000004038       45000.00               0  
1000007540        5000.00               0  
1000011046       19500.00               0

**Podemos hacernos preguntas para verificar hipótesis:**

**1)** ¿Hay proyectos sin patrocinadores pero con plata prometida?

In [75]:
kickstarter_2018[(kickstarter_2018.backers == 0) & (kickstarter_2018.usd_pledged > 0)]

name  \
ID                                                              
100173829   La Press: Make Soap at Home in 2 minutes. No C...   
1004972393  AlwaysRemembered.com The new way to remember o...   
1006945199                     Have your Cake and Eat it too!   
1019661678                            Presidential Pareidolia   
1031663961                              The Lite Bros. Series   
1037887032            The Undergraduate Scholarship Quarterly   
1039895327                                            iSchool   
1039940191                                  XAM?N: 100% AGAVE   
104562843                      Royalty-Free Trap Music Videos   
1050338924           Families Who Eat Together Stay Together!   
1051154899  Undocumented: Funds for publishing of my ficti...   
1051720010                                   Peperless Future   
1052301248                                          Blood Ash   
1064127689                      Courage to Conquer Web Series   
107462950   Help @Vita to record an album to come back his...   
1088207503             OWR Roots. Arts. Culture Festival 2017   
1097673976                                Cheering from Italy   
1097782445                                        Paws2People   
1097822147                                 "PIÉNSALO 2 VECES"   
1103822371                Patches for everyone - Café Viereck   
110453253   HOME Emergency Preparedness. Resiliency. Safet...   
1104637995                                    Harry's DogPack   
1130613074  Luce Quadra : Coffee table, glove box & ambien...   
1136802303                                            BOULDER   
1137626463  Rigid3D, Medium to Large High Quality Precisio...   
1138397473                               THE SNEAKER CLEANERS   
1138745536  "Born Again" - Share Your Inner Beauty with th...   
1140686007  SILVER PAGES + LIFELINES (double project relea...   
1150466424  Want to play the board game KICKSTARTER YOU AR...   
116438579   100% Silk Necklaces & Bracelet, Silk Thread Ta...   
...                                                       ...   
771293950      Pixelate: The Next Game to Destroy Friendships   
773493220   Grabación de Albúm España 2017 Señor Serio - S...   
775139653                           The Park Pay and Stay App   
778477372                  AWESOME WORLD : DISCOVER THE WORLD   
783060249                             Treasures From The Word   
786348291                20s A Difficult Age - The Video Game   
797562974                                             You-Con   
80485427    BusinessTerra Worldwide Platform affordable fo...   
805836463   Funding a Performing Arts Theatre for Children...   
806472586                                  A Game of Soldiers   
816152387                                     Good Earth Cafe   
817495998                                         3d Memories   
831864842                                         Alpha Prime   
831982875                                        Timeline App   
841884816      All Natural Bath & Body Treats Beauty Gift Box   
846555805   Vintage Skate-Vase, Limited Edition of 250, Nu...   
871593952                Programmieren für ein besseres Leben   
888587273                          Fatty-D's Charburger Grill   
89103022    Innovative exercise equipment developed by Sym...   
904996877                 Campus Food Ordering App (FOODORME)   
909080977                End Of The Trail Records Kickstarter   
911925118        International/Domestic Student room platform   
913066071                           Snack Gloves and Glovkins   
916919369                         The Last of the Blood God's   
927798302                            Randezy's first Mixtape!   
928930112   Hitchhike for 5,610 miles and try to reach Ell...   
930939140                   RetroNemesis Arcade Clothing Line   
945359364                     Painting the faces of strangers   
950767056                                       Sins Of A Son   
958530198   DIY Easy To Han

* Tenemos 315 proyectos sin patrocinadores pero con plata prometida.

**2)** ¿Hay proyectos no exitosos que pasaron el objetivo?

In [76]:
kickstarter_2018[(kickstarter_2018.state != 'successful') & (kickstarter_2018.pledged > kickstarter_2018.goal)]

name        category  \
ID                                                                              
100149523                         Duncan Woods - Chameleon EP           Music   
1001732009  DotPen : Redefines drawing and writing for sma...  Product Design   
1002379225  The World's First Genuine, Intuitively-control...      Technology   
1002854072  Spycup turns your iPhone into a hidden camera ...         Gadgets   
1003023003          The Making of Ashley Kelley's Debut Album           Music   
1003130892                       Butter Side Down Debut Album           Music   
1003500859  I Can't Believe It's Not Potato Salad ! (Cance...       Cookbooks   
1003629045                            Chase Goehring debut EP           Music   
100372771   A Big Dream of a Little Man - Alex Oddman's De...           Music   
1004013077           Spencer Capier Instrumental Project 2015           Music   
1004126342  LUKAS LIGETI'S 50TH BIRTHDAY FESTIVAL: ORIGINA...           Music   
1004255433      The Battle For Breukelen: A Neighborhood Epic    Film & Video   
1004298993  "Tamboura Plays Violin" - a collection of Pop ...           Music   
1004402863        Nightingale Noel - An A Cappella Holiday CD           Music   
1004434782  DIY BOXCAT by Rato Kim - The Vinyl Art Toy Pla...             DIY   
1005618996      Masters of the Tooniverse Podcast (Suspended)       Webseries   
1006327667                           Letters to the Wild - EP           Music   
100822204   Freedom Ride Jacket | Revolutionizing Urban Cy...  Product Design   
1008675685               Inspired & the Sleep EP - Eyelid Kid           Music   
1008815806  Into Winter: An Instrumental Christmas Experiment           Music   
1009260867  Rude Story Cards. An easy, fun & outrageous pa...  Tabletop Games   
1009774693                Universal Peace | Universal Justice           Music   
1010217550                               Solarboot-Havel-Tour           Music   
1010694326                                          Surrender           Music   
1011508551                               EBENEZER Solo Record           Music   
1012704793                               Life In The Bus Lane    Film & Video   
1012807897                                  Juliet Remembered    Film & Video   
1014077376   Joseph in the Well's Debut EP - Indie String Pop           Music   
1015045443                   Support Nadine's Musical Venture           Music   
1015771770                                           THE POET    Film & Video   
...                                                       ...             ...   
981103283   Chappie OST Listening Party 12 Hr #TheChappeni...   Installations   
981146646                      Dollar Bill Origami (Canceled)             Art   
981182186                           Save the Ramzi (Canceled)           Music   
982327172   Kalimba The Spirit of Earth Wind and Fire DVD ...           Music   
982329957              The Atlanta Vocal Project Christmas CD           Music   
982721469             TAVAPUTS Vinyl Record "Unresting Event"           Music   
983679756    Custom Hand-made 3-Strand Bracelets! (Suspended)         Fashion   
983728691                    The Octopodes' 5th Studio Album!           Music   
984297189                    Major Arcana Tarot by A MIYAKO M    Illustration   
985231591                                            TWO FISH    Film & Video   
986616604                   What awaits in Austin (Suspended)      Photobooks   
988763792     Maud Sings Maud: A Musical Betsy-Tacy Companion           Music   
991409513                           Chris Copeland's First EP           Music   
991483230   Support Vocal Synergy's Second Semester and Te...           Music   
991820324                                   The Blend CD 2015           Music   
992433603                                   Jonny Come Lately    Film & Video   
992987490   Bicycle Neon Card - Save the night of old Hong...   Playing Cards   
9935913

* Tenemos 2952 proyectos no exitosos que pasaron el objetivo

**3)** ¿Cómo es, en promedio, la correlación entre el objetivo y la plata obtenida?

In [77]:
df = kickstarter_2018.groupby('state')[['goal','pledged']].aggregate('mean')
df['pledged'] / df['goal']

state
canceled     0.03
failed       0.02
live         0.12
successful   2.37
suspended    0.06
undefined    0.52
dtype: float64

* En promedio, entre el objetivo y la plata obtenida, la correlación es positiva

**4)** ¿Cuánta plata se obtiene en promedio por estado y categoría?

In [78]:
kickstarter_2018.pivot_table(values='usd_pledged',index='state',columns='main_category')

main_category     Art  Comics  Crafts   Dance   Design  Fashion  Film & Video  \
state                                                                           
canceled       405.94  722.46  342.51  455.18  3361.22   996.49       1468.05   
failed         554.10  869.11  263.54  555.21  2353.08   686.00       1368.40   
live           887.49  934.78   95.43  182.59  3559.01   443.48        336.93   
successful    5420.99 9178.68 4053.92 4445.79 43545.99 15136.85      12403.29   
suspended      436.70  347.47  201.33  320.43 11471.42  3539.29       2463.24   
undefined         nan     nan     nan    0.00     0.00     0.00          0.00   

main_category     Food    Games  Journalism   Music  Photography  Publishing  \
state                                                                          
canceled        715.49  3312.50      198.61  585.91       418.70      451.11   
failed          927.01  1970.92      452.95  557.37       565.53      539.22   
live           1014.11  1678.80      762.78  434.78       832.48     1031.71   
successful    14227.41 41643.05     8598.27 6283.11      7897.06     7257.91   
suspended      2708.65  2497.19       44.77  340.77       159.96      422.83   
undefined          nan     0.00         nan    0.00         0.00        0.00   

main_category  Technology  Theater  
state                               
canceled          3959.58  1472.81  
failed            2018.73   892.50  
live              2028.49   479.18  
successful       67226.64  5289.17  
suspended        22953.52    57.08  
undefined            0.00     0.00

### Ejercicio 5

**1)** ¿Hay proyectos éxitosos que no consiguieron el objetivo? Si hay, ¿Qué porcentaje sí y cuál no?    

In [79]:
# Proyectos exitosos que NO consiguieron el objetivo

Exitoso_noObjetivo= (kickstarter_2018[(kickstarter_2018.state == 'successful') & (kickstarter_2018.pledged < kickstarter_2018.goal)]).name
Exitoso_noObjetivo

ID
1368992939                                         Santa's Sled
1769772834    Borders of the Imagination - An Epic Sci-Fi/Fa...
251325270                     Documentary film about Les Blank.
4732285         Insiders/Out: Exploring Outsider Art in America
780863434              TUW Ruhla Watches - Comeback of a Legend
Name: name, dtype: object

In [80]:
# Proyectos exitosos que SI consiguieron el objetivo

Exitoso_siObjetivo= (kickstarter_2018[(kickstarter_2018.state == 'successful') & (kickstarter_2018.pledged >= kickstarter_2018.goal)]).name
Exitoso_siObjetivo

ID
1000014025                                 Monarch Espresso Bar
1000023410    Support Solar Roasted Coffee & Green Energy!  ...
100005484                                      Lisa Lim New CD!
1000057089    Tombstone: Old West tabletop game and miniatur...
1000070642                  Mike Corey's Darkness & Light Album
1000072011                CMUK. Shoes: Take on Life Feet First.
1000091520                          The Book Zoo - A Mini-Comic
1000102741    Matt Cavenaugh & Jenny Powers make their 1st a...
1000104688                                  Permaculture Skills
1000104953    Rebel Army Origins: The Heroic Story Of Major ...
100011318                         My Moon - Animated Short Film
1000117861                    Ledr workbook: one tough journal!
1000120151              Feather Cast Furled Fly Fishing Leaders
100012079               Chris Eger Band - New Nashville Record!
1000131947                           Arrows & Sound Debut Album
1000132945    Zen in America: a film 

In [81]:
total= kickstarter_2018.name
total.shape

(378657,)

In [82]:
Exitoso_noObjetivo.shape

(5,)

In [83]:
Exitoso_siObjetivo.shape

(133951,)

In [84]:
# Porcentaje de proyectos exitosos que NO consiguieron el objetivo sobre el total
porcentaje_no= (Exitoso_noObjetivo.count()/ total.count())
porcentaje_no

1.3204562440414412e-05

In [85]:
# Porcentaje de proyectos exitosos que SI consiguieron el objetivo sobre el total
porcentaje_si= (Exitoso_siObjetivo.count()/ total.count())
porcentaje_si

0.35375286869119016

In [86]:
# Proporción de los proyectos exitosos que SI consiguieron el objetivo sobre los que NO lo consiguieron
proporcion_1= Exitoso_siObjetivo.count()/(Exitoso_siObjetivo.count() + Exitoso_noObjetivo.count())
proporcion_1

0.9999626743109677

In [87]:
# Proporción de los proyectos exitosos que NO consiguieron el objetivo sobre los que SI lo consiguieron
proporcion_2= Exitoso_noObjetivo.count()/(Exitoso_noObjetivo.count() + Exitoso_siObjetivo.count())
proporcion_2

3.732568903221953e-05

**2)** Calcular una tabla con la cantidad de proyectos por categoría principal y estado.

In [95]:
kickstarter_2018.pivot_table(index='main_category',
                             columns='state',
                             aggfunc='count', 
                             fill_value=0).astype(int)

backers                                            category  \
state         canceled failed live successful suspended undefined canceled   
main_category                                                                
Art               2222  14130  194      11510        96         0     2222   
Comics             842   4036   76       5842        23         0      842   
Crafts             843   5703   76       2115        72         0      843   
Dance              163   1235   18       2338        13         1      163   
Design            4152  14814  305      10550       246         2     4152   
Fashion           2650  14182  250       5593       138         3     2650   
Film & Video      5755  32903  332      23623       117       854     5755   
Food              2211  15969  184       6085       153         0     2211   
Games             6202  16002  287      12518       220         1     6202   
Journalism         523   3137   31       1012        52         0      523   
Music             3305  21752  281      24197       149      2234     3305   
Photography        986   6384   48       3305        55         1      986   
Publishing        3602  23145  299      12300        66       462     3602   
Technology        4715  20616  377       6434       424         3     4715   
Theater            608   3708   41       6534        21         1      608   

                                        ...    usd_pledged             \
state         failed live successful    ...           live successful   
main_category                           ...                             
Art            14130  194      11510    ...            194      11510   
Comics          4036   76       5842    ...             76       5842   
Crafts          5703   76       2115    ...             76       2115   
Dance           1235   18       2338    ...             18       2338   
Design         14814  305      10550    ...            305      10550   
Fashion        14182  250       5593    ...            250       5593   
Film & Video   32903  332      23623    ...            332      23623   
Food           15969  184       6085    ...            184       6085   
Games          16002  287      12518    ...            287      12518   
Journalism      3137   31       1012    ...             31       1012   
Music          21752  281      24197    ...            281      24197   
Photography     6384   48       3305    ...             48       3305   
Publishing     23145  299      12300    ...            299      12300   
Technology     20616  377       6434    ...            377       6434   
Theater         3708   41       6534    ...             41       6534   

                                  usd_pledged_real                         \
state         suspended undefined         canceled failed live successful   
main_category                                                               
Art                  96         0             2222  14130  194      11510   
Comics               23         0              842   4036   76       5842   
Crafts               72         0              843   5703   76       2115   
Dance                13         1              163   1235   18       2338   
Design              246         2             4152  14814  305      10550   
Fashion             138         3             2650  14182  250       5593   
Film & Video        117       854             5755  32903  332      23623   
Food                153         0             2211  15969  184       6085   
Games               220         1             6202  16002  287      12518   
Journalism           52         0              523   3137   31       1012   
Music               149      2234             3305  21752  281      24197   
Photography          55         1              986   6384   48       3305   
Publishing           66       462             3602  23145  299      12300   
Technology          424         3             4715  20616  377 

## 2.6. Identificar y documentar valores atípicos/outliers


Queremos analizar las características de los proyectos que obtienen dinero pero sin considerar los casos atípicos.

In [ ]:
%matplotlib inline
kickstarter_2018.usd_pledged.hist();

In [ ]:
kickstarter_2018.boxplot(column='usd_pledged');

"There are many rules of thumb for how to define an outlier but there is no silver bullet. One rule of thumb is that any data point that is three standard deviations away from the mean of the same data point for all observations."

In [ ]:
# descartamos los valores que están a 3 valores de la desviación estandar.
# +- pero aca no tenemos valores negativos

outliers = kickstarter_2018[kickstarter_2018.usd_pledged > (kickstarter_2018.usd_pledged.mean() + 
                                                            3 * kickstarter_2018.usd_pledged.std())]
outliers

In [ ]:
kickstarter_2018.drop(outliers.index).boxplot(column='usd_pledged');

In [ ]:
kickstarter_2018.drop(outliers.index).boxplot(column='usd_pledged',by='state');

Analizar outliers por categorías

In [ ]:
kickstarter_2018.boxplot(column='usd_pledged',by='state');
kickstarter_2018.boxplot(column='usd_goal_real',by='state');

#### Ejercicio 6

Calcular los valores atípicos de 'usd_goal_real' y graficar los boxplots, con y sin estos valores por categoría

## 2.7. Evaluar cómo comprimir los datos 
para su almacenamiento más eficiente.

En la actualidad, en las nubes el almacenamiento no suele ser problema sino la velocidad para leerlo que suele ser determinante a la hora de determinar el formato de almacenamiento.


## 2.8. Guardar el set de datos con un nombre informativo.


EL problema de la ingeniería de software

-----------------------------------------------------------------